### DISCUSSION TITLE
How to include interactive effects in inlabru

### DISCUSSION THREAD

**Discussion by wanxinru** [inlabru-org/inlabru#153] [Q&A] [answered]
Prof. Finn,

Many thanks for your previous help, I tried to include the interaction term following the form of "comp = ~ ... + interaction(~ -1 + x1:x2, model = "fixed")" but with following error:

Error in duplicated.default(temp, fromLast = fromLast, ...) : 
 duplicated() applies only to vectors

enclosed with the data the code:

Regards,
Xinru

> fit <- bru(components =
>                 y ~ x1 + x2 + x3 + 
>                 interaction(~ -1 + x1:x2, model = "fixed"), 
>               family = "zeroinflatednbinomial1", 
>               data = data_sp)
> 
> Error in duplicated.default(temp, fromLast = fromLast, ...) : 
>   duplicated() applies only to vectors

> data_sp
> class       : SpatialPointsDataFrame 
> features    : 4498 
> extent      : 74.60803, 132.3199, 19.16057, 52.22981  (xmin, xmax, ymin, ymax)
> crs         : NA 
> variables   : 4
> names       :    y,                x1,                x2,                x3 
> min values  :    0, -10.1445992177037, -20.7533789311727, -24.8418413889503 
> max values  : 1531,  23.4976168588322,  37.2222060302636,  27.7707840598426 
> 
> head(data_sp)
>    y        x1       x2       x3
> 1  5 -4.833604 3.133297 15.11054
> 2 77  5.196780 3.133297 15.11054
> 3 15  3.845258 3.133297 15.11054
> 4 12  3.020332 3.133297 15.11054
> 5 16 -1.871995 3.133297 15.11054
> 6  7  1.159075 3.133297 15.11054

**Comment by finnlindgren** [comment, marked as answer]
Can you post a complete code example (including data construction etc)?

**Reply by wanxinru** [reply]
Hi, here is the data and code, fit1 without interactive effects run well, but fit2 doesn't, hope you can run it, thanks for your time.

[data_0711.csv](https://github.com/inlabru-org/inlabru/files/9462296/data_0711.csv)

```
data <- read_csv("dataframe/data_0711.csv")
data <- data[!is.na(data$lat)&!is.na(data$lon), ] %>% as.data.frame()
data_sp <- SpatialPointsDataFrame(data[,c("lon", "lat")], data)

fit1 <- bru(components =
              cases ~ rod_div + rod_diviucn + prec,
            family = "zeroinflatednbinomial1",
            data = data_sp)

fit2 <- bru(components =
              cases ~ rod_div + rod_diviucn + prec +
              interaction(~ -1 + rod_div:rod_diviucn, model = "fixed"),
            family = "zeroinflatednbinomial1",
            data = data_sp)
```

**Reply by finnlindgren** [reply]
What's your `R` and `inlabru` versions? Your example runs fine here.   If you have the CRAN version of inlabru, try updating to the github development version and test it again.
```
library(inlabru)
library(tidyverse)

data <- read_csv("data_0711.csv")
data <- data[!is.na(data$lat)&!is.na(data$lon), ] %>% as.data.frame()
data_sp <- SpatialPointsDataFrame(data[,c("lon", "lat")], data)

fit1 <- bru(components =
              cases ~ rod_div + rod_diviucn + prec,
            family = "zeroinflatednbinomial1",
            data = data_sp)

fit2 <- bru(components =
              cases ~ rod_div + rod_diviucn + prec +
              interaction(~ -1 + rod_div:rod_diviucn, model = "fixed"),
            family = "zeroinflatednbinomial1",
            data = data_sp)

bind_rows(fit2$summary.fixed,
          fit2$summary.random$interaction)

#                     mean          sd  0.025quant     0.5quant   0.975quant         mode          kld                  ID
# rod_div      7.631006983 2.364561021  3.25669590  7.533208323 12.556294275  7.342550438 3.508829e-06                <NA>
# rod_diviucn  0.465493106 0.203130824  0.03267647  0.477484454  0.831816094  0.501311515 5.381128e-06                <NA>
# prec         0.003886396 0.002712211 -0.00125053  0.003818303  0.009411852  0.003687844 1.283603e-06                <NA>
# Intercept   -3.783140582 2.599826061 -8.39667343 -3.964598690  1.824246025 -4.327512276 6.229831e-06                <NA>
# ...5        -0.485917906 0.175003048 -0.85138859 -0.478338051 -0.163081202 -0.463532427 3.745222e-06 rod_div:rod_diviucn
```

**Reply by finnlindgren** [reply]
It may be a bit clearer to include all the simple fixed effects in a single component:
```
fit3 <- bru(components =
              cases ~ -1 +
              interaction(~ 1 + prec + rod_div*rod_diviucn, model = "fixed"),
            family = "zeroinflatednbinomial1",
            data = data_sp)

fit3$summary.random$interaction

#                    ID         mean          sd   0.025quant     0.5quant   0.975quant         mode          kld
# 1         (Intercept) -3.783158901 2.599533968 -8.396237076 -3.964566130  1.823479136 -4.327408111 6.229124e-06
# 2                prec  0.003886317 0.002711893 -0.001249949  0.003818223  0.009411111  0.003687751 1.283120e-06
# 3             rod_div  7.631041725 2.364381233  3.257061098  7.533255528 12.555912577  7.342605168 3.508182e-06
# 4         rod_diviucn  0.465506448 0.203109800  0.032741909  0.477494956  0.831793524  0.501317928 5.380798e-06
# 5 rod_div:rod_diviucn -0.485919265 0.174989806 -0.851360115 -0.478340025 -0.163107926 -0.463534456 3.744909e-06
```

**Reply by wanxinru** [reply]
Thanks for your suggestions, here is the information on R and inlabru versions:

> R version 4.2.1 (2022-06-23 ucrt) -- "Funny-Looking Kid"
> 
> Package: inlabru
> Type: Package
> Title: Bayesian Latent Gaussian Modelling using INLA and Extensions
> Version: 2.5.2

I reinstalled inlabru with the development version (v. 2.5.2.9000) and everything is ok, run the same results as you.

Then I tried to continue the analysis to visualize the interactive effects between rod_div and rod_diviucn, I tried to predict the fitted cases by the data frame with rod_div and rod_diviucn each running from 1:100 in every unique combination, while the other variables were their means. But when I continue predicting, there is the following error:

> Error in unique.default(x, nmax = nmax) : 
> unique() applies only to vectors


Here is the original code:
```
data_pred = expand.grid(rod_div = seq(from = min(data[,"rod_div"], na.rm = T), to = max(data[,"rod_div"], na.rm = T), length.out = 20), 
                        rod_diviucn = seq(from = min(data[,"rod_diviucn"], na.rm = T), to = max(data[,"rod_diviucn"], na.rm = T), length.out = 20))

data_pred = cbind(data_pred, t(as.data.frame(colMeans(data[, c("year", "lat", "lon", "cases", 
                                                               "temp", "prec", "popd", "NDVI", "GDP", "elev")], na.rm = T))))

data_pred_sp <- sp::SpatialPointsDataFrame(data_pred[,c("lon", "lat")], data_pred)

results <- predict(object = fit3, data_pred_sp, 
                     formula = ~ interaction(~ 1 + prec + rod_div*rod_diviucn, model = "fixed"))
```


When I'm writing this reply, I try using an easier formula: ~ rod_div*rod_diviucn, and surprisingly get the results:


```
results <- predict(object = fit3, data_pred_sp, 
                     formula = ~ rod_div*rod_diviucn) 

> head(results$mean)
[1] -35.68878 -32.82145 -29.95412 -27.08680 -24.21947 -21.35214
```



I'm not sure if it is the correct way, hopefully, yes.

**Reply by finnlindgren** [reply]
No, neither prediction expression are what you need. The `formula` argument to `inlabru::predict()` just defines a generak R expression, so this is incorrect syntax:
```
results <- predict(object = fit3, data_pred_sp, 
                     formula = ~ interaction(~ 1 + prec + rod_div*rod_diviucn, model = "fixed"))
```
Your other version is a valid R expression, but it just evaluates the product between the covariate values; not the estimated effects.
You need something like this instead:
```
results <- predict(object = fit3, data_pred_sp, 
                     formula = ~ interaction) 
```
That should compute the combined effect of all the "interaction" components.
If you need to access the coefficients, you can get it with `interaction_latent`. Something like this:
```
results <- predict(object = fit3, data_pred_sp, 
                     formula = ~ interaction_latent) 
```
There isn't currently an easy way to compute just part of a `model="fixed"` component effect. You need to find the relevant latent index and extract it from the latent vector like this:
```
idx <- which(fit3$summary.random$interaction$ID == "rod_div:rod_diviucn")
res2 <- predict(object = fit3, data_pred_sp, 
                formula = ~ interaction_latent[idx] * (rod_div * rod_diviucn) 
)                   

```

**Reply by wanxinru** [reply]
Many thanks for your help, I tried to predict the interactive effects following `formula = ~ interaction`, there are quite some differences with the results using `formula = ~ rod_div*rod_diviucn`, indicating it is an important option to get a precise prediction. 

I think there is enough information on how to include interactive effects in inlabru in this post, this step is VERY important for my current study, thank you again for your help and patience.

**Reply by finnlindgren** [reply]
No, `formula = ~ rod_div*rod_diviucn` evaluates the actual product between the values of the covariates, and doesn't involve the estimated model at all, so you should not be comparing those two outputs.

**Reply by finnlindgren** [reply]
The _only_ place where the R "formula syntax" is used, is in the input to the interaction component definition.

**Reply by finnlindgren** [reply]
And the only reason the `like()` and `predict()` functions call the predictor expressions "formula" is for backwards compatibility for models that only involve plain fixed effect covariate terms. A better name would be "predictor".

**Reply by wanxinru** [reply]
I see, `formula = ~ rod_div*rod_diviucn` is not a matter of accuracy, it's wrong, and I didn't expect that the "formula" here played the role of "predictor".
